spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoHierarchySpec.groovy

```diff
@@ -0,0 +1,94 @@
+package org.spockframework.runtime.model
+
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class SpecInfoHierarchySpec extends Specification {
+
+  @Shared def grandparent = new SpecInfo()
+  @Shared def parent = new SpecInfo()
+  @Shared def specInfo = new SpecInfo()
+  @Shared def child = new SpecInfo()
+  @Shared def grandchild = new SpecInfo()
+
+  @Shared hierarchy = [grandparent, parent, specInfo, child, grandchild]
+  @Shared hierarchyNames = ["grandparent", "parent", "specInfo", "child", "grandchild"]
+
+  def setupSpec() {
+    parent.superSpec = grandparent
+    grandparent.subSpec = parent
+
+    specInfo.superSpec = parent
+    parent.subSpec = specInfo
+
+    child.superSpec = specInfo
+    specInfo.subSpec = child
+
+    grandchild.superSpec = child
+    child.subSpec = grandchild
+  }
+
+  @Unroll("#name")
+  def "top spec"() {
+    expect:
+    underTest.getTopSpec() == grandparent
+
+    where:
+    underTest << hierarchy
+    name << hierarchyNames
+  }
+
+  @Unroll("#name")
+  def "bottom spec"() {
+    expect:
+    underTest.getBottomSpec() == grandchild
+
+    where:
+    underTest << hierarchy
+    name << hierarchyNames
+  }
+
+  @Unroll("#name")
+  def "specs top to bottom"() {
+    expect:
+    underTest.specsTopToBottom == hierarchy
+
+    where:
+    underTest << hierarchy
+    name << hierarchyNames
+  }
+
+  @Unroll("#name")
+  def "specs bottom to top"() {
+    expect:
+    underTest.specsBottomToTop == hierarchy.reverse()
+
+    where:
+    underTest << hierarchy
+    name << hierarchyNames
+  }
+
+  @Unroll("#name")
+  def "specs to bottom"() {
+    expect:
+    underTest.specsToBottom == hierarchy.subList(hierarchy.indexOf(underTest), hierarchy.size())
```

Comment:
Isn't this a tautological test, you are basically repeating the implementation here. If there was an error in the logic it wouldn't be caught as it uses the same logic.